### PR TITLE
Make generated client extend generated trait

### PR DESCRIPTION
**Pull request by ignasi35**
Before this PR, the client side of a gRPC service was:

```scala
final class ShoppingServiceClient extends ShoppingService with AkkaGrpcClient{

  // the impl for ShoppingService methods
  // the impl for AkkaGrpcClient methods

  // lifted methods for ShoppingService client
}
```

This PR introduces `ShoppingServiceLiftedClient` and repurposes `ShoppingServiceClient`:

after:

```scala
trait ShoppingServiceClient 
     extends ShoppingServiceLiftedClient 
     with    ShoppingService 
     with    AkkaGrpcClient

trait ShoppingServiceLiftedClient {
  // (noop) lifted methods for ShoppingService client
}

final class RawShoppingServiceClient{
  // the impl for ShoppingService methods
  // the impl for AkkaGrpcClient methods

  // (actual impl) lifted methods for ShoppingService client
}
```

This separation, including a `noop` implementation of the lifted methods in `ShoppingServiceLiftedClient` gives some extra flexibility to users:

 * users can stub `ShoppingServiceClient` (before the `final` made it impossible)
 * users can stub `ShoppingServiceClient` knowing there's no Netty/gRPC machinery under the hood.
 * production code using a client may depend on `ShoppingServiceLiftedClient` or `ShoppingService` to limit the usage of the power API (before, `ShoppingServiceClient` exposed all giving too much power).

-------

This is a new approach to #442

### CODE REVIEW

**Comment by dwijnand** [comment]
Rebased to clean up the diff/review.

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt [hunk 1]
```diff
@@ -91,3 +99,19 @@ object @{service.name}Client {
         .build()
   }
 }
+
+
+trait @{service.name}LiftedClient {
```
Comment (raboof) [inline codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: While we've been speaking about 'lifted', I'm not sure I actually like calling it that. What do you think about `@{service.name}ClientPowerApi` or something similar?

File: codegen/src/main/twirl/templates/JavaClient/Client.scala.txt [hunk 2]
```diff
@@ -36,31 +36,30 @@ import scala.concurrent.Future;
...
-  private final GrpcClientSettings settings;
-  private final CallOptions options;
-  private final Materializer mat;
-  private final ExecutionContext ec;
+public abstract class @{service.name}Client extends @{service.name}LiftedClient implements @{service.name}, AkkaGrpcClient {
+  public static final @{service.name}Client create(GrpcClientSettings settings, Materializer mat, ExecutionContext ec) {
+    return new Raw@{service.name}Client(settings, mat, ec);
+  }
 
+  protected final static class Raw@{service.name}Client extends @{service.name}Client {
```
Comment (raboof) [inline codegen/src/main/twirl/templates/JavaClient/Client.scala.txt]: You called it `Default` before - that sounded a bit less scary than 'raw'. Why the change?

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt [hunk 3]
```diff
@@ -91,3 +99,19 @@ object @{service.name}Client {
...
+  @for(method <- service.methods) {
+    /**
+     * Lower level "lifted" version of the method, giving access to request metadata etc.
+     * prefer @{method.name}(@method.parameterType) if possible.
+     */
+    @if(method.methodType == akka.grpc.gen.Unary || method.methodType == akka.grpc.gen.ClientStreaming) {
+      def @{method.name}(): SingleResponseRequestBuilder[@method.parameterType, @method.outputTypeUnboxed] = ???
+    } else {
+      def @{method.name}(): StreamResponseRequestBuilder[@method.parameterType, @method.outputTypeUnboxed] = ???
```
[the remark this answers is not in the document]
Comment (ignasi35) [reply codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: Counldn't agree more. :-)

File: codegen/src/main/twirl/templates/JavaClient/Client.scala.txt (hunk 2 above)
Comment (ignasi35) [reply codegen/src/main/twirl/templates/JavaClient/Client.scala.txt]: I've juggled three different names as I toyed with ideas. No preference for any of them. `Default` is good.

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt [hunk 4]
```diff
@@ -18,8 +18,19 @@ import scala.concurrent.{ ExecutionContext, Future, Promise }
...
-final class @{service.name}Client(settings: GrpcClientSettings)(implicit mat: Materializer, ex: ExecutionContext) extends @{service.name} with AkkaGrpcClient {
-  import @{service.name}Client._
+// Not sealed so users can extend to write their stubs
+trait @{service.name}Client extends @{service.name} with @{service.name}ClientPowerApi with AkkaGrpcClient
+
+object @{service.name}Client {
+  def apply(settings: GrpcClientSettings)(implicit mat: Materializer, ex: ExecutionContext): @{service.name}Client =
+    new Default@{service.name}Client(settings)
+  // A factory method use by some tooling to reflectively create a @{service.name}Client.
+  def create(settings: GrpcClientSettings, mat: Materializer, ex: ExecutionContext): @{service.name}Client =
```
Comment (raboof) [inline codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: Hmm so it's not possible to discover the apply method in the same way? interesting

Comment (ignasi35) [reply codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: TBH, I haven't tried. I first attempted to make it work consolidating the `XService.create(settings, mat, ex)` idiom for both scala and java.

But it works indeed. I'll simplify...

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt [hunk 5]
```diff
@@ -22,11 +22,9 @@ import io.grpc._
 trait @{service.name}Client extends @{service.name} with @{service.name}ClientPowerApi with AkkaGrpcClient
 
 object @{service.name}Client {
+  // Note, apply is also use by some tooling to reflectively create a @{service.name}Client.
```
Comment (raboof) [inline codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: I'm not sure if this is worth commenting here as we have tests catching it

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ignasi35) [reply codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: https://github.com/akka/akka-grpc/issues/463

File: runtime/src/main/scala/akka/grpc/internal/AkkaGrpcClientFactory.scala [hunk 6]
```diff
@@ -6,19 +6,21 @@ package akka.grpc.internal
...
 
 object AkkaGrpcClientFactory {
   def create[T <: AkkaGrpcClient: ClassTag](settings: GrpcClientSettings)(implicit mat: Materializer, ex: ExecutionContext): T = {
-    classTag[T].runtimeClass.asInstanceOf[Class[T]]
-      .getConstructor(
-        classOf[GrpcClientSettings],
-        classOf[Materializer],
-        classOf[ExecutionContext])
-      .newInstance(settings, mat, ex)
+    // this reflection requires:
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/internal/AkkaGrpcClientFactory.scala]: This is no longer true... perhaps remove comment? when someone is this deep in the internals already, I guess it's not such a big step to get it from the code anymore? Plus this is expected to change again later.
